crosshair/libimpl/builtinslib.py

```diff
@@ -1929,13 +1929,39 @@ def __getitem__(self, i):
         return SmtStr(smt_result)
 
     def find(self, substr, start=None, end=None):
-        smt_mystr = self.var
-        smt_substr = force_to_smt_sort(substr, SmtStr)
-        if end is not None:
-            end = force_to_smt_sort(end, SmtInt)
-            smt_mystr = z3.SubString(smt_mystr, 0, end)
-        start = 0 if start is None else force_to_smt_sort(start, SmtInt)
-        return SmtInt(z3.IndexOf(smt_mystr, smt_substr, start))
+        start = 0 if start is None else start + len(self) if start < 0 else start
+        if end is not None and end <= start:
+            return -1
+        value = self[slice(start, end, 1)].var
+        sub = force_to_smt_sort(substr, SmtStr)
+        if self.statespace.smt_fork(z3.Contains(value, sub)):
+            return SmtInt(z3.IndexOf(value, sub, 0) + start)
+        else:
+            return -1
+
+    def rfind(self, substr, start=None, end=None):
+        start = 0 if start is None else start + len(self) if start < 0 else start
+        if end is not None and end <= start:
+            return -1
```

Comment:
At first I thought that the slice would take care of this for us?

Then I noticed this, which made me think some sort of special handling might be necessary. Sigh!:
```
>>> 'abcd'[2:1].find('')
0
>>> 'abcd'.find('', 2, 1)
-1
```

```

Reply:
Yup, noticed the same thing, which is why that line is in there. I think we've reached the "right" implementation. However, the fuzz test still finds issues. How do I find out what inputs are causing it?

Reply:
I want to refactor that to use pytest.parametrize, but for now, this is what I do:
```
python crosshair/fuzz_core_test.py FuzzTest.test_[whatever] -v 2>&1 >log.txt
```
And search log.txt for several asterisks to find to the failing example.

Note that when exceptions are thrown, we try to be consistent with the exception type, but the exception message doesn't have to match.

Reply:
It's finding some obviously pathological input which complicates things a bit, but that's ducktyping for ya. Funnily enough it also didn't trigger when making the slice object, but only when trying to use it.